### PR TITLE
Feat: refactoring for adding run-all support

### PR DESCRIPTION
**Pull request by TomerHeber**
### Solution

Some refactoring I wanted to do before working on #438.
(Instead of having one large PR).

Schema validation was moved to the correct place, instead of doing it before the API call.
Updated tests accordingly.

### CODE REVIEW

File: client/template.go [hunk 1]
```diff
@@ -108,19 +109,33 @@ type VariablesFromRepositoryPayload struct {
...
 	if payload.Type != "terragrunt" && payload.TerragruntVersion != "" {
-		return Template{}, errors.New("can't define terragrunt version for non-terragrunt blueprint")
+		return errors.New("can't define terragrunt version for non-terragrunt template")
 	}
 	if payload.Type == "terragrunt" && payload.TerragruntVersion == "" {
-		return Template{}, errors.New("must supply Terragrunt version")
+		return errors.New("must supply terragrunt version")
+	}
+
+	if payload.Type == "cloudformation" && payload.FileName == "" {
```
Comment (TomerHeber) [inline client/template.go:128]: moved this validation from a different place (this is the correct place).

File: client/template.go [hunk 2]
```diff
@@ -108,19 +109,33 @@ type VariablesFromRepositoryPayload struct {
 	Repository           string   `json:"repository"`
 }
 
-func (client *ApiClient) TemplateCreate(payload TemplateCreatePayload) (Template, error) {
```
Comment (TomerHeber) [inline client/template.go:111]: function not removed. diff mix-up.

File: client/template.go [hunk 3]
```diff
@@ -149,18 +164,6 @@ func (client *ApiClient) TemplateDelete(id string) error {
 }
 
 func (client *ApiClient) TemplateUpdate(id string, payload TemplateCreatePayload) (Template, error) {
-	if payload.Name == "" {
```
Comment (TomerHeber) [inline client/template.go:152]: validation is done in the schema code and not in the API code.

File: client/template_test.go [hunk 4]
```diff
@@ -220,59 +220,6 @@ var _ = Describe("Templates Client", func() {
 		})
 	})
 
-	Describe("Terragrunt version errors", func() {
```
Comment (TomerHeber) [inline client/template_test.go:223]: new tests added to the schema.

File: env0/resource_template.go [hunk 5]
```diff
@@ -51,14 +51,8 @@ func templateCreatePayloadFromParameters(d *schema.ResourceData) (client.Templat
 	templateCreatePayloadRetryOnHelper(d, "deploy", &payload.Retry.OnDeploy)
 	templateCreatePayloadRetryOnHelper(d, "destroy", &payload.Retry.OnDestroy)
 
-	if payload.Type == "cloudformation" {
```
Comment (TomerHeber) [inline env0/resource_template.go:54]: added this to the above "validate" function.

File: env0/resource_template_test.go [hunk 6]
```diff
@@ -57,6 +57,7 @@ func TestUnitTemplateResource(t *testing.T) {
 			},
 		},
 		Type:               "terragrunt",
+		TerragruntVersion:  "0.35.1",
```
Comment (TomerHeber) [inline env0/resource_template_test.go:60]: now the tests are working better. (They were supposed to fail without terragrunt_version, but didn't)

**Review by yaronya — APPROVED** [review]

**Pull request merged**